rfd/xxxx-config-resource-docs.md

```diff
@@ -0,0 +1,554 @@
+functions, `analysis.Analyzer`s. However, these static analysis tools are
+intended for linting, not generating reference documentation. 
+
+While you can configure an `analysis.Analyzer` to return a value from its `Run`
+function, the `multichecker` and `singlechecker` drivers run by exposing a
+`Main` function that handles all of the program's output and CLI arguments. It
+is not possible to initialize a data structure outside the `Main` function and
+assign values to it during the `Run` function, since `Main`
+[exits](https://cs.opensource.google/go/x/tools/+/refs/tags/v0.8.0:go/analysis/multichecker/multichecker.go;l=59)
+the current process after completing, leaving the program no chance to process
+the data structure.
+
+The `singlechecker` and `multichecker` packages run static analysis by importing
+from a `checker` package, which is internal to the
+`analysis` package. While we could copy this code and
+modify it for our purposes, the reference generator does not require the
+general-purpose static analysis capabilities of the `analysis` package.
+
+### Converting structs to `Resource`s
+
+In Go's `ast` package, a [`Decl`](https://pkg.go.dev/go/ast#Decl) is an
+interface representing a declaration. The logic for generating a `Resource`
+from a given type declaration depends on the nature of the type declaration. 
+
+Start by asserting that the `Decl` is an `*ast.GenDecl` with a `Specs`
+containing a `TypeSpec` with `Type: StructType`. From there, we can map the
+fields within [`ast.StructType`](https://pkg.go.dev/go/ast#StructType) to fields
+in a new `Resource`.
+
+Each `StructType` has a [`*FieldList`](https://pkg.go.dev/go/ast#FieldList) that
+we can use to obtain field data, including comments and tags.
+
+|`Resource` Field|How to generate|
+|---|---|
+|`Resource.SourcePath`|The `file` property of the underlying `sourceDecl`.|
+|`Resource.Description`|`GenDecl.Doc`|
+|`Resource.SectionName`|`TypeSpec.Name.Name`, with camelcase converted to separate words.|
+|`Fields[*].Name`|Process the `Field.Tag` in the `FieldList`, extracting the value of the `json` tag.|
+|`Fields[*].Description`|Each [`Field`](https://pkg.go.dev/go/ast#Field) in a `FieldList` has a `*CommentGroup` that we can use to extract the field's GoDoc. If a comment begins with the name of the field, we can remove the field name.|
+|`Fields[*].Type`|We can use `Field.Type` within the `FieldList` for this. Follow the rules for processing field types [below](#processing-field-types) |
+|`YAMLExample`|Follow the rules for processing field types [below](#processing-field-types).|
+
+If a root type named within the config file is not a struct, the generator will
+exit with an error message.
+
+### Processing field types
+
+#### Custom fields
+
+Entries in the reference for custom fields, e.g., `Labels`, don't lend
+themselves to automatic generation because they implement custom unmarshalers,
+and we can't rely on `json` struct tags to indicate the types of these fields.
```

Comment:
Could we make an educated guess and provide a way to override it? So simple numeric/string fields are taken from the Go type, as well as references to other structs?

Edit: I think that's what you explained below.

Reply:
I'll add my response to this in my response to [this comment](https://github.com/gravitational/teleport/pull/25823#discussion_r1191510099).